src/shared/copy.c

```diff
@@ -794,6 +797,68 @@ static int prepare_nocow(int fdf, const char *from, int fdt, unsigned *chattr_ma
         return 0;
 }
 
+/* Copies fs-verity status.  May re-open fdt to do its job. */
+static int copy_fs_verity(int fdf, int *fdt) {
```

Comment:
hmm, i find it a bit weird to look at the verity status of the source file. Wouldn't it make more sense to just globally enable verity as we copy, regardless of the fsvreity status of the source?

i mean, i think we should be able to build fsverity images from any build tree, regardless if that build tree is on btrfs of xfs or tmpfs or weirdfs, and i don't really see a reason why the resulting image should look any different.

we care a lot about reproducibility, hence I think it would be really wise if we'd minimize how much of the source file system feature set leaks into the final image.

Hence, dumb this down, make it a simple boolean you can enable in [Partition] that just dumbly enables verity for all files it drops in, regardless of the source state?

Reply:
Or are you concerned that that would enable verity for too many files, where you don't want that? is that a real problem? 

if it is, maybe make this an enum: FSVerity=yes|no|propagate or so, to always enable, always disable or to propagate the fsverity state.

Reply:
This isn't how composefs (or ostree) works in practice: we have fs-verity enabled files in the object store but they sit in the same partition alongside the mutable state of `/etc` or `/var` of the OS deployments.

So yes: we very much do want this done on a per-file basis.

Reply:
here's another idea: maybe in addition to CopyFiles= have CopyFilesVerity= which would be just like CopyFiles= but also enable verity. Thus you could compose your fs from a bunch of files to copy in with verity and a couple without, and the result would be a merged tree?

Reply:
I think adding an optional flags field to 'CopyFiles=` is the way to go. You'd have `source:target:flags` and one of the flags would be `fsverity`

Reply:
> Or are you concerned that that would enable verity for too many files, where you don't want that? is that a real problem?

Yeah that would definitely be a problem, you don't want to enable this everywhere, it wouldn't make much sense, as it's largely filesystem and use case dependent.

I think the current approach is best, where the metadata is copied if present. If you prep a directory tree with this metadata and then pass it as source to report, that is indication enough that you want it to be in the dest img too, I don't think we need to bother with flags, we already have so many

Reply:
> I think the current approach is best, where the metadata is copied if present. If you prep a directory tree with this metadata and then pass it as source to report, that is indication enough that you want it to be in the dest img too, I don't think we need to bother with flags, we already have so many

There's only one problem, then: `mke2fs` requires an explicit `-O verity` to create a filesystem with fs-verity support.  The patch I'm writing there won't even query the status of the files on disk if this option isn't enabled.  So we'd still need to figure out when we pass that option to `mkfs.ext4` (this is not an issue with btrfs).

So it leads us back to the question about when we should pass `-O verity` to `mkfs.ext4`.  I don't know an (already existing) way to smuggle a "we saw at least one file with fs-verity enabled" bit out of the recursive tree copy operation, and even if there was an easy way, it would be a bit magic...

Thoughts welcome.

Reply:
That's not a problem, we support env vars to pass options to mkfs, eg:

`export SYSTEMD_REPART_MKFS_OPTIONS_EXT4="-O verity"`

Reply:
by making this explicit via the flags it's also very clear whether `-O verity` needs to be specified during mkfs: if there's one CopyFiles= request with verity=on or verity=copy then we enable it, otherwise we don't.

Reply:
> by making this explicit via the flags it's also very clear whether -O verity needs to be specified during mkfs: if there's one CopyFiles= request with verity=on or verity=copy then we enable it, otherwise we don't.

This is definitely an argument in favour of this approach.

> after all fsverity is mostly redundant metadata, it can be generated on-the-fly if missing. It's hence mostly a boolean, not more.

jein... there's also the choice of which algorithm, blocksize, and salt to use, and signatures (if we ever wanted to support that).  Of course, everyone I know is using blocksize of 4096, sha256 and no salt, without signatures... but we already have a couple of hypothetical situations in mind where the block size might change, as an example.  For me, the most useful way to indicate that would be by setting the parameters when enabling fs-verity on the source files via ` FS_IOC_ENABLE_VERITY`.

Is there a known use-case for a filesystem where all files have fs-verity enabled?

Reply:
I've done what I suggested.  It would be possible to add `fsverity=always` in the future, if we cared to address the issues raised.